src/simulator/halo_simulator.py

```diff
@@ -0,0 +1,217 @@
+# Copyright 2021 The Private Cardinality Estimation Framework Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Simulates the Halo system.
+
+This code simulates the part of the Halo system from the Measurement
+Coordinator downwards.  This includes the SMPC and the publishers.
+The Halo Simulator represents the functionality and interface
+that is assumed to exist by the modeling strategies that are examined as
+part of this evaluation effort.
+"""
+
+from typing import List
+from typing import Set
+from typing import Tuple
+
+import numpy as np
+
+from wfa_cardinality_estimation_evaluation_framework.estimators.same_key_aggregator import (
+    StandardizedHistogramEstimator,
+)
+from wfa_cardinality_estimation_evaluation_framework.common.noisers import (
+    LaplaceMechanism,
+)
+from wfa_cardinality_estimation_evaluation_framework.estimators.vector_of_counts import (
+    VectorOfCounts,
+)
+
+from wfa_planning_evaluation_framework.data_generators.data_set import DataSet
+from wfa_planning_evaluation_framework.models.reach_curve import ReachCurve
+from wfa_planning_evaluation_framework.models.reach_point import ReachPoint
+from wfa_planning_evaluation_framework.simulator.privacy_tracker import (
+    DP_NOISE_MECHANISM_GAUSSIAN,
+)
+from wfa_planning_evaluation_framework.simulator.privacy_tracker import NoisingEvent
+from wfa_planning_evaluation_framework.simulator.privacy_tracker import PrivacyBudget
+from wfa_planning_evaluation_framework.simulator.privacy_tracker import PrivacyTracker
+from wfa_planning_evaluation_framework.simulator.publisher import Publisher
+from wfa_planning_evaluation_framework.simulator.simulation_parameters import (
+    SimulationParameters,
+)
+
+
+class HaloSimulator:
+    """Simulator for the Halo System.
+
+    A Halo Simulator simulates the Halo System from the Measurement Coordinator
+    downwards.  In particular, it simulates the SMPC and the behavior of the
+    publishers.  The Halo Simulator represents the functionality and interface
+    that is assumed to exist by the modeling strategies that are examined as
+    part of this evaluation effort.
```

Comment:
Nice description!

To clarify the relationship between `HaloSimulator` and `PlannerSimulator`, how about adding something like
"Explicitly, `HaloSimulator` simulates (i) the observable, DP data points and (ii) the DP single-pub reach curves that will be later used for fitting the multi-pub reach surface in `PlannerSimulator`."

Reply:
Done.